### ISSUE TITLE
kubectl follow logs stops after few seconds if there is a lot data to stdout

### ISSUE THREAD

**Issue by avneradania1** [kubernetes/kubectl#1548]
### What happened?

A lot of running jobs generates a massive amount of log to stdout.
Steps:
1. Running K8s job that invokes a lot of lines to stdout.
2. kubectl logs -f <podName> -n <nameSpace>
Starts tailing pod output, till it stuck randomly. 
**Note**, it stuck till pod exit with a success or failure. 
 

### What did you expect to happen?

Expected result is to see all pod stdout on screen. This is not happening. 

### How can we reproduce it (as minimally and precisely as possible)?

1. invokes a pod that generates a massive amount of stdout.
2. run:
   # kubectl logs -f <pod-name> -n <namespace>
3. after a while, this is randomly duration, it stops showing output 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.7
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
$ uname -a
# Linux <hostname> 3.10.0-1160.99.1.el7.x86_64 kubernetes/kubernetes#1 SMP Wed Sep 13 14:19:20 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux



</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by mpuckett159** [comment]
When logs are tailed kubectl uses the watcher utility, which relies on go channels, in this case a buffered go channel. We can see in the below link that the default size for the buffer is 100, and this is not currently configurable. Now, normally this would be fine because the logs coming in would be sent out, but because you are generating so much data, the logs are unable to exit the channel buffer at a faster rate than the logs are flowing in to the channel buffer, resulting in negative pressure and the channel to hang.

[How the log function spawns multiple watchers](https://github.com/kubernetes/kubernetes/blob/00c30941260a27e6929aef84c7fdbc8f1508518c/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L346-L345)

[The default buffer size of the channel used by the watchers](https://github.com/kubernetes/kubernetes/blob/7e434682e450e28d36f0ee4787e7b4672e8eb255/staging/src/k8s.io/apimachinery/pkg/watch/watch.go#L51-L53)

If you run the command causing the issue with `-v=9` flag that may give us a clearer indication that this is what is happening. Could you do that and post the resulting logs for that message (not the logs that are being output but the ones being returned as a result of the verbosity flag)?

**Comment by avneradania1** [comment]
@mpuckett159,  Thanks for detailed response. I'm guessing you're asking for the beginning of v=9 output. 
if so, file attached with details.
> 
[kubectl-v9.pdf](https://github.com/kubernetes/kubectl/files/14123665/kubectl-v9.pdf)

**Comment by Pawan-Bishnoi** [comment]
would like to help here.

Do we want to provide an option to increase the buffer size?

**Comment by Pawan-Bishnoi** [comment]
not stale

**Comment by Pawan-Bishnoi** [comment]
@mpuckett159 does the logs shared above confirm your theory?

I can see that [DefaultConsumeRequest](https://github.com/kubernetes/kubernetes/blob/00c30941260a27e6929aef84c7fdbc8f1508518c/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L413-L421) uses the [stream](https://github.com/kubernetes/client-go/blob/690fd5274dfc971dcda653e45bc6e6852ab57316/rest/request.go#L966-L970) function of go-client.

I don't see watch package used in this code path. What am I missing? 🤔

**Comment by Pawan-Bishnoi** [comment]
I am surprised that more people aren't complaining about this 😄 
This blocks me daily and I would love to help solve whatever limitation we are running into here.

**Comment by avneradania** [comment]
Agree, it looks like people does not look into stdout anymore, but on main Log servers such as Loki, Elk etc.

**Comment by ardaguclu** [comment]
I spent some time on this today and it was hard to reproduce. I think, checking the kubelet and api server logs if that happens can enlighten the issue. I'm not sure this issue is definitely happening on kubectl and we need to eliminate the other components first.

**Comment by ardaguclu** [comment]
Also another question would be; does this happen without tailing `-f`?

**Comment by Pawan-Bishnoi** [comment]
I doubt kubelet too.
From whatever I could see, it wasn't happening on kubectl. Will dig some more on my next attempt.

**Comment by Pawan-Bishnoi** [comment]
not stale

**Comment by Pawan-Bishnoi** [comment]
not stale

**Comment by mpuckett159** [comment]
I think we still need more information on this, we don't have a root cause of this issue yet, so we need to track that down first.

That being said, if the issue is what I think it is (the channel being overwhelmed with data) I don't really know how we would fix this, and I think that my recommendation would be that we don't pursue fixing it because if we have an edge case where the rate of log data being written to a channel is faster than the CPU can read from the channel, this is a pretty extreme edge case, and if you are trying to get logs from this kind of service you should be using a more robust logging solution than kubectl.

**Comment by brianpursley** [comment]
> 2\. **Note**, it stuck till pod exit with a success or failure.

I'm starting to think this might be a container runtime issue, or possibly kubelet as others have mentioned.

Either way, let's try to troubleshoot this a little more and then transfer this issue to the right people, if it's not something that can be fixed in kubectl.

@avneradania @Pawan-Bishnoi 
When this happens, if you run `kubectl logs` again or from another process, does it work?  Or will that second attempt also be "stuck" right away too?  I'm trying to understand if the logs for a pod are permanently broken for the rest of its lifetime, or if restarting the kubectl command allows them to work again.

Also, can you provide the container runtime that your cluster is using, like "containerd://1.7.18" in the following output from `kubectl get node -o wide`
```
~ $ kubectl get node -o wide
NAME                 STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
kind-control-plane   Ready    control-plane   51d   v1.31.2   172.18.0.2    <none>        Debian GNU/Linux 12 (bookworm)   6.8.0-57-generic   containerd://1.7.18
```

**Comment by Pawan-Bishnoi** [comment]
> or if restarting the kubectl command allows them to work again

Yes, this was the case earlier. Kubectl logs when run again would work for a brief amount of time.
If I reduced log volume by changing log level etc then it would never block. Only when log volume was high, it used to get stuck soon after I hit enter.

Since it seems to be _gone_ now, I am ok closing it 😄
